File: jars/src/org/niatahl/tahlan/skills/DaemonicWarfare.java
package org.niatahl.tahlan.skills;

import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.characters.MarketSkillEffect;
import com.fs.starfarer.api.impl.campaign.ids.Stats;

public class DaemonicWarfare {

    public static float FLEET_SIZE = 50f;
    public static int DEFEND_BONUS = 50;


    public static class Level1 implements MarketSkillEffect {
        public void apply(MarketAPI market, String id, float level) {
            market.getStats().getDynamic().getMod(Stats.COMBAT_FLEET_SIZE_MULT).modifyFlat(id, FLEET_SIZE / 100f, "Daemonic Warfare");
        }

        public void unapply(MarketAPI market, String id) {
            market.getStats().getDynamic().getMod(Stats.COMBAT_FLEET_SIZE_MULT).unmodifyFlat(id);
        }

        public String getEffectDescription(float level) {
            return "+" + (int) Math.round(FLEET_SIZE) + "% fleet size";
        }

        public String getEffectPerLevelDescription() {
            return null;
        }

        public ScopeDescription getScopeDescription() {
            return ScopeDescription.GOVERNED_OUTPOST;
        }
    }

    public static class Level2 implements MarketSkillEffect {
        public void apply(MarketAPI market, String id, float level) {
            market.getStats().getDynamic().getMod(Stats.GROUND_DEFENSES_MOD).modifyMult(id, 1f + DEFEND_BONUS * 0.01f, "Daemonic Warfare");
        }

        public void unapply(MarketAPI market, String id) {
            //market.getStats().getDynamic().getMod(Stats.GROUND_DEFENSES_MOD).unmodifyPercent(id);
            market.getStats().getDynamic().getMod(Stats.GROUND_DEFENSES_MOD).unmodifyMult(id);
        }

        public String getEffectDescription(float level) {
            return "+" + (int) (DEFEND_BONUS) + "% effectiveness of ground defenses";
        }

        public String getEffectPerLevelDescription() {
            return null;
        }

        public ScopeDescription getScopeDescription() {
            return ScopeDescription.GOVERNED_OUTPOST;
        }
    }

}

File: jars/src/org/niatahl/tahlan/skills/DaemonicWarfare.kt
package org.niatahl.tahlan.skills

import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.characters.LevelBasedEffect.ScopeDescription
import com.fs.starfarer.api.characters.MarketSkillEffect
import com.fs.starfarer.api.impl.campaign.ids.Stats
import kotlin.math.roundToInt

object DaemonicWarfare {

    const val FLEET_SIZE = 50f
    const val DEFEND_BONUS = 50

    class Level1 : MarketSkillEffect {
        override fun apply(market: MarketAPI, id: String, level: Float) {
            market.stats.dynamic.getMod(Stats.COMBAT_FLEET_SIZE_MULT)
                .modifyFlat(id, FLEET_SIZE / 100f, "Daemonic Warfare")
        }

        override fun unapply(market: MarketAPI, id: String) {
            market.stats.dynamic.getMod(Stats.COMBAT_FLEET_SIZE_MULT).unmodifyFlat(id)
        }

        override fun getEffectDescription(level: Float): String = "+${FLEET_SIZE.roundToInt()}% fleet size"

        override fun getEffectPerLevelDescription(): String? = null

        override fun getScopeDescription(): ScopeDescription = ScopeDescription.GOVERNED_OUTPOST
    }

    class Level2 : MarketSkillEffect {
        override fun apply(market: MarketAPI, id: String, level: Float) {
            market.stats.dynamic.getMod(Stats.GROUND_DEFENSES_MOD)
                .modifyMult(id, 1f + DEFEND_BONUS * 0.01f, "Daemonic Warfare")
        }

        override fun unapply(market: MarketAPI, id: String) {
            market.stats.dynamic.getMod(Stats.GROUND_DEFENSES_MOD).unmodifyMult(id)
        }

        override fun getEffectDescription(level: Float): String = "+$DEFEND_BONUS% effectiveness of ground defenses"

        override fun getEffectPerLevelDescription(): String? = null

        override fun getScopeDescription(): ScopeDescription = ScopeDescription.GOVERNED_OUTPOST
    }
}
